benchmark/Benchmark.h

```diff
@@ -0,0 +1,250 @@
+// Copyright 2022, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Andre Schlegel (November of 2022, schlegea@informatik.uni-freiburg.de)
+
+#pragma once
+
+#include <vector>
+#include <functional>
+#include <string>
+#include <optional>
+
+#include "util/json.h"
+#include "util/Timer.h"
+#include <util/HashMap.h>
+#include <util/Exception.h>
+#include "../benchmark/util/HashMapWithInsertionOrder.h"
+
+/*
+ * Used for measuring the time needed for the execution of a function. It also
+ * saves the measured time togehter with a description of the function
+ * measured.
+ */
+class BenchmarkRecords {
+
+  public:
+
+    // Describes a measured function.
+    struct RecordEntry {
+      std::string descriptor_; // Needed, because without it, nobody could tell,
+                              // which time belongs to which benchmark.
+      float measuredTime_; // The measured time in seconds.
+
+    // JSON (d)serialization for all the created structures in BenchmarkRecords.
+    NLOHMANN_DEFINE_TYPE_INTRUSIVE(BenchmarkRecords::RecordEntry, descriptor_, measuredTime_)
+    };
+
+    // Describes a group of measured functions.
+    struct RecordGroup {
+      std::string descriptor_; // Needed for identifying groups.
+      std::vector<RecordEntry> entries_;
+
+    // JSON (d)serialization for all the created structures in BenchmarkRecords.
+    NLOHMANN_DEFINE_TYPE_INTRUSIVE(BenchmarkRecords::RecordGroup, descriptor_, entries_)
+    };
+    
+    // Describes a table of measured functions.
+    struct RecordTable {
+      // For identification.
+      std::string descriptor_;
+      // The names of the columns and rows.
+      std::vector<std::string> rowNames_;
+      std::vector<std::string> columnNames_;
+      // The time measurements in seconds. Access is [row, column].
+      std::vector<std::vector<std::optional<float>>> entries_;
+
+      RecordTable(const std::string& pDescriptor, 
+          const std::vector<std::string>& pRowNames,
+          const std::vector<std::string>& pColumnNames):
+          descriptor_{pDescriptor}, rowNames_{pRowNames},
+          columnNames_{pColumnNames},
+          entries_(pRowNames.size(), std::vector<std::optional<float>>(pColumnNames.size())) {}
+
+      // JSON (d)serialization for all the created structures in BenchmarkRecords.
+      NLOHMANN_DEFINE_TYPE_INTRUSIVE(BenchmarkRecords::RecordTable, descriptor_, rowNames_, columnNames_, entries_)
+    };
```

Comment:
Add the option, to manually set table entries, so that helpful information can be added inside the table. Maybe make the entries a variant of std::optional<float> and std::string?

Reply:
Done, see [7019669](https://github.com/ad-freiburg/qlever/pull/860/commits/70196692c25671279c4bacfbd407aaab4600e015).